src/open_inwoner/accounts/tests/test_auth_2fa_sms.py

```diff
@@ -0,0 +1,542 @@
+            response = self.client.post(
+                reverse("login"),
+                data={
+                    "next": "",
+                    "username": self.user.email,
+                    "password": "secret",
+                },
+                follow=True,
+            )
+
+            params = {
+                "next": "",
+                "user": signer.sign(self.user.pk),
+            }
+            self.assertRedirects(
+                response, f"{reverse('verify_token')}?{urlencode(params)}"
+            )
+            self.assertContains(response, "Account verificatie")
+            self.assertTrue(mock_gateway_send.called)
+            self.assertEqual(mock_gateway_send.call_count, 27)
+            self.assertEqual(
+                mock_gateway_send.call_args[1]["to"], self.user.phonenumber
+            )
+
+            # Unblocked attempt.
+            location = f"{reverse('verify_token')}?{urlencode(params)}"
+            response = self.client.post(
+                path=f"{reverse('resend_token')}?{urlencode(params)}",
+                follow=True,
+                HTTP_REFERER=location,
+            )
+            self.assertEqual(mock_gateway_send.call_count, 28)
+            self.assertEqual(
+                mock_gateway_send.call_args[1]["to"], self.user.phonenumber
+            )
+            self.assertRedirects(
+                response, f"{reverse('verify_token')}?{urlencode(params)}"
+            )
+
+    @override_settings(ACCOUNTS_USER_TOKEN_EXPIRE_TIME=150)
+    @patch("open_inwoner.accounts.gateways.gateway.send")
+    def test_login_resend_sms_user_token_expires(self, mock_gateway_send):
+        now = timezone.now()
+        some_time_ago = now - datetime.timedelta(seconds=151)
+
+        with dj_freeze_time(time.mktime(some_time_ago.timetuple())):
+            params = {
+                "next": "",
+                "user": signer.sign(self.user.pk),
+            }
+
+        location = reverse("verify_token") + "?" + urlencode(params)
+        response = self.client.post(
+            reverse("resend_token") + "?" + urlencode(params),
+            follow=True,
+            HTTP_REFERER=location,
+        )
+        self.assertEqual(mock_gateway_send.call_count, 0)
+        self.assertRedirects(response, str(settings.LOGIN_URL))
+
+    @patch("open_inwoner.accounts.gateways.gateway.send")
+    def test_login_by_adding_phone_number_succeeds(self, mock_gateway_send):
+        self.user.phonenumber = ""
+        self.user.save()
+
+        response = self.client.post(
+            reverse("login"),
+            data={
+                "next": "",
+                "username": self.user.email,
+                "password": "secret",
+            },
+            follow=True,
+        )
+        self.assertContains(response, "Account verificatie (stap 1 van 2)")
+        self.assertFalse(mock_gateway_send.called)
+
+        response = self.client.post(
+            response._request.get_full_path(),
```

Comment:
This feels weird:

````
    response = self.client.post(
            response._request.get_full_path(),
````

Maybe just explicitly use the real url? Now there is indirection and we have to search and think to read it.

Reply:
The first response redirects the user to the phonenumber view so we have to use this in order to be aware of the url with the query params.